🌱 test: add dynamic finalizer assertions for e2e framework

adityabhatia opened the pull request:
<!-- Thanks for sending a pull request! Here are some tips for you:
    1. If this is your first time, please read our contributor guidelines: https://github.com/kubernetes-sigs/cluster-api/blob/main/CONTRIBUTING.md#contributing-a-patch and developer guide https://github.com/kubernetes-sigs/cluster-api/blob/main/docs/book/src/developer/guide.md

    2. Please add an icon to the title of this PR (see https://sigs.k8s.io/cluster-api/CONTRIBUTING.md#contributing-a-patch), and delete this line and similar ones
    the icon will be either ⚠️ (:warning:, major or breaking changes), ✨ (:sparkles:, feature additions), 🐛 (:bug:, patch and bugfixes), 📖 (:book:, documentation or proposals), or 🌱 (:seedling:, minor or other) 
-->

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes # https://github.com/kubernetes-sigs/cluster-api/issues/9969

<!-- 
Please label this pull request according to what area(s) you are addressing. For reference on PR/issue labels, see: https://github.com/kubernetes-sigs/cluster-api/labels?q=area+

Area example:
/area runtime-sdk
-->
/area testing

sbueringer commented:
/ok-to-test

sbueringer commented:
@adityabhatia There seems to be something wrong with the commit. The diff looks correct in GitHub but based on the commit title and locally it is a merged commit.

sbueringer commented on test/framework/finalizers_helpers.go:
```diff
@@ -111,6 +128,8 @@ func getObjectsWithFinalizers(ctx context.Context, proxy ClusterProxy, namespace
 
 		if len(setFinalizers) > 0 {
+			// assert if the expected finalizers are set on the resource
+			Expect(allFinalizerAssertions[node.Object.Kind]).To(Equal(setFinalizers), "for resource type ", node.Object.Kind)
```
first parameter should be actual

sbueringer commented on test/framework/finalizers_helpers.go:
```diff
@@ -149,6 +168,18 @@ func assertFinalizersExist(ctx context.Context, proxy ClusterProxy, namespace st
+	allFinalizerAssertions := make(map[string][]string, 0)
+	for i := range finalizerAssertions {
+		for kind, finalizers := range finalizerAssertions[i] {
+			allFinalizerAssertions[kind] = finalizers
```
Maybe we should throw an error if we find duplicate assertions for the same kind

adityabhatia replied:
True, I added aggregated errors for such cases.

sbueringer commented on test/framework/finalizers_helpers.go:
```diff
@@ -106,6 +124,8 @@ func getObjectsWithFinalizers(ctx context.Context, proxy ClusterProxy, namespace
 
 		if len(setFinalizers) > 0 {
+			// assert if the expected finalizers are set on the resource
+			Expect(setFinalizers).To(Equal(allFinalizerAssertions[node.Object.Kind]), "for resource type ", node.Object.Kind)
```
```suggestion
			Expect(setFinalizers).To(Equal(allFinalizerAssertions[node.Object.Kind]), "for resource type %s", node.Object.Kind)
```

optionalDescription is handled like this for len >= 2
```go
return fmt.Sprintf(optionalDescription[0].(string), optionalDescription[1:]...) + "\n"
```

adityabhatia replied:
Always thought optional arguments are concatenated

sbueringer commented on test/framework/finalizers_helpers.go:
```diff
@@ -143,3 +163,24 @@ func assertFinalizersExist(ctx context.Context, proxy ClusterProxy, namespace st
+	for i := range finalizerAssertions {
+		for kind, finalizers := range finalizerAssertions[i] {
+			if _, alreadyExists := allFinalizerAssertions[kind]; alreadyExists {
+				allErrs = append(allErrs, fmt.Errorf("finaliser assertion cannot be applied as it already exists for kind: %s, existing value: %v, new value: %v",
```
```suggestion
				allErrs = append(allErrs, fmt.Errorf("finalizer assertion cannot be applied as it already exists for kind: %s, existing value: %v, new value: %v",
```

sbueringer commented:
@adityabhatia Happy to merge this PR now that we reverted the other PR that broke CI (+/- addressing nits / rebase). To unblock the work in CAPV>

adityabhatia commented:
> @adityabhatia Happy to merge this PR now that we reverted the other PR that broke CI (+/- addressing nits / rebase)., to unblock the work in CAPV.

Rebased on main (post revert) and ready for review :)

sbueringer commented:
Thx!

/lgtm
/approve

sbueringer commented:
/cherry-pick release-1.6
